### PR TITLE
HIVE-21614: "Support" comparing Derby's table parameter CLOB by replacing `=` with `LIKE`

### PR DESCRIPTION
**Pull request by hankfanchiu**
### What changes were proposed in this pull request?

On the Hive MetaStore server's side of the `HiveMetaStoreClient#listTableNamesByFilter()` API, conditionally manipulate the string filter by replacing any `=` comparison on a table parameter with a `LIKE` operator.

This replacement is only performed if:

1. The database product backing the Hive MetaStore is Derby,
2. The `=` comparison is on a table parameter, i.e. the `"TABLE_PARAMS"."PARAM_VALUE"` column.

### Why are the changes needed?

HIVE-12274 changed the type of the `"TABLE_PARAMS"."PARAM_VALUE"` column to `CLOB` for Derby and Oracle, e.g.:

https://github.com/apache/hive/blob/7d4134e7fe9bfb8b8aa3344a9ae72f4f36c98b2c/metastore/scripts/upgrade/derby/039-HIVE-12274.derby.sql#L8-L12

With the type change, invoking `ObjectStore#listTableNamesByFilter()` given a `=` comparison -- on a table parameter stored in Derby -- fails with the following exception:

```
ERROR 42818: Comparisons between 'CLOB (UCS_BASIC)' and 'CLOB (UCS_BASIC)' are not supported. Types must be comparable. String types must also have matching collation. If collation does not match, a possible solution is to cast operands to force them to the default collation (e.g. SELECT tablename FROM sys.systables WHERE CAST(tablename AS VARCHAR(128)) = 'T1')
```

Without reverting the column type change, we add support for `=` comparison on table parameters in Derby. This addresses HIVE-21614.

### Does this PR introduce _any_ user-facing change?

Yes, invoking `HiveMetaStoreClient#listTableNamesByFilter()` with a `$param_key = '$param_value'` filter now succeeds for Derby databases.

### CODE REVIEW

**Comment by hankfanchiu** [comment]
@pvary, mind taking a look?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java [hunk 1]
```diff
@@ -4693,6 +4694,17 @@ private Table ensureGetTable(String catName, String dbName, String tblName)
     return convertToTable(ensureGetMTable(catName, dbName, tblName));
   }
 
+  /**
+   * Conditionally alters an {@link ExpressionTree} to be compatible with certain database products.
+   */
+  private void reconcileTreeWithDatabaseProduct(ExpressionTree tree) throws MetaException {
+    if (sqlGenerator.getDbProduct().isDERBY()) {
```
Comment (pvary) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:4701]: Is this only problem for Derby?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/OperatorEqualsToLikeReplacer.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * Replaces any {@link Operator.EQUALS} with {@link Operator.LIKE} for {@link LeafNode}
+ * representing an equality comparison on a table parameter.
+ * 
+ * This utility is effective for database products that do not support equality comparisons
+ * between values of the CLOB type.
+ */
+final class OperatorEqualsToLikeReplacer extends TreeVisitor {
+  @Override
+  protected void visit(LeafNode node) throws MetaException {
+    if (
```
Comment (pvary) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/OperatorEqualsToLikeReplacer.java:37]: nit: Could we format this way:
```
      if (node.keyName.startsWith(hive_metastoreConstants.HIVE_FILTER_FIELD_PARAMS)
          && node.operator == Operator.EQUALS) {
```

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java (hunk 1 above)
Comment (hankfanchiu) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:4701]: @findepi mentioned in https://issues.apache.org/jira/browse/HIVE-21614?focusedCommentId=16849571&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16849571 that Oracle may be implicated as well, but I have not verified.

Having this `ObjectStore#reconcileTreeWithDatabaseProduct` defined allows us to conveniently add other databases, though, so perhaps we could revisit as needed?

**Comment by hankfanchiu** [comment]
@pvary, is this good to go?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java (hunk 1 above)
Comment (pvary) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java:4701]: There are Docker images for the supported databases. Could you please run tests with them?

**Comment by dengzhhu653** [comment]
Hi @hankfanchiu, sorry for the inconvenience. Are you still working on this issue now?

**Comment by hankfanchiu** [comment]
> Are you still working on this issue now?

@dengzhhu653, I am not.

**Pull request closed without merging**